images/controller/internal/controllers/rvr_diskful_count/reconciler.go

```diff
@@ -18,39 +18,324 @@ package rvrdiskfulcount
+// getDiskfulReplicaCountFromReplicatedStorageClass gets the diskful replica count based on ReplicatedStorageClass.
+//
+// If replication = None, returns 1; if replication = Availability, returns 2;
+// if replication = ConsistencyAndAvailability, returns 3.
+func getDiskfulReplicaCountFromReplicatedStorageClass(ctx context.Context, cl client.Client, rv *v1alpha3.ReplicatedVolume) (int, error) {
+	// Get ReplicatedStorageClass name from ReplicatedVolume
+	rscName := rv.Spec.ReplicatedStorageClassName
+	if rscName == "" {
+		return 0, fmt.Errorf("ReplicatedVolume has empty ReplicatedStorageClassName")
+	}
+
+	// Get ReplicatedStorageClass object
+	rsc := &v1alpha1.ReplicatedStorageClass{}
+	err := cl.Get(ctx, client.ObjectKey{Name: rscName}, rsc)
+	if err != nil {
+		return 0, fmt.Errorf("getting ReplicatedStorageClass %s: %w", rscName, err)
+	}
+
+	// Determine diskful replica count based on replication
+	switch rsc.Spec.Replication {
+	case v1alpha3.ReplicationNone:
+		return 1, nil
+	case v1alpha3.ReplicationAvailability:
+		return 2, nil
+	case v1alpha3.ReplicationConsistencyAndAvailability:
+		return 3, nil
 	default:
-		r.log.Error("unknown req type", "typedReq", typedReq)
-		return reconcile.Result{}, e.ErrNotImplemented
+		return 0, fmt.Errorf("unknown replication value: %s", rsc.Spec.Replication)
+	}
+}
+
+// getDiskfulReplicatedVolumeReplicas gets all Diskful ReplicatedVolumeReplica objects for the given ReplicatedVolume
+// by the spec.replicatedVolumeName and spec.type fields. Returns a map with RVR name as key and RVR object as value.
+// Returns empty map if no RVRs are found.
+func getDiskfulReplicatedVolumeReplicas(ctx context.Context, cl client.Client, rv *v1alpha3.ReplicatedVolume) (map[string]*v1alpha3.ReplicatedVolumeReplica, error) {
+	allRvrList := &v1alpha3.ReplicatedVolumeReplicaList{}
+	err := cl.List(ctx, allRvrList)
+	if err != nil {
+		return nil, fmt.Errorf("listing all ReplicatedVolumeReplicas: %w", err)
+	}
+
+	// Filter by spec.replicatedVolumeName and build map
+	rvrMap := make(map[string]*v1alpha3.ReplicatedVolumeReplica)
+
+	for i := range allRvrList.Items {
+		if allRvrList.Items[i].Spec.ReplicatedVolumeName == rv.Name && allRvrList.Items[i].Spec.Type == v1alpha3.ReplicaTypeDiskful {
+			rvrMap[allRvrList.Items[i].Name] = &allRvrList.Items[i]
+		}
+	}
+
+	return rvrMap, nil
+}
+
+// splitReplicasByDeletionStatus splits replicas into two maps: one with replicas that have DeletionTimestamp,
+// and another with replicas that don't have DeletionTimestamp.
+// Returns two maps with RVR name as key and RVR object as value. Returns empty maps if no RVRs are found.
+func splitReplicasByDeletionStatus(totalRvrMap map[string]*v1alpha3.ReplicatedVolumeReplica) (deletedRvrMap, nonDeletedRvrMap map[string]*v1alpha3.ReplicatedVolumeReplica) {
+	deletedRvrMap = make(map[string]*v1alpha3.ReplicatedVolumeReplica)
+	nonDeletedRvrMap = make(map[string]*v1alpha3.ReplicatedVolumeReplica)
```

Comment:
Я бы память сразу выделил под них, пусть и немного больше

```suggestion
	deletedRvrMap = make(map[string]*v1alpha3.ReplicatedVolumeReplica, len(totalRvrMap))
	nonDeletedRvrMap = make(map[string]*v1alpha3.ReplicatedVolumeReplica, len(totalRvrMap))
```